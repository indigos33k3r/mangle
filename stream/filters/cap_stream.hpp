#ifndef MANGLE_STREAM_CAP_H
#define MANGLE_STREAM_CAP_H

#include "../stream.hpp"
#include <stdexcept>

namespace Mangle {
namespace Stream {

/** A read filter that limits the total readable number of bytes to a
    given amount. Returns eof() after that.

    This can be used as a simplified alternative to slice_stream that:
    - doesn't need or support seeking
    - doesn't support pointer operations
    - is read-only

    It is perfect for sequential reading of archive files, where you
    just want to read a part of a file once, especially if the outer
    stream is compressed and doesn't support seeking.

    The stream has size() and tell() which reflect the internal size
    and position within the cap stream, starting at zero and ending at
    size().

    Premature eof()s in the source stream, ie. running out of data
    before 'size' has been reached, will cause an exception to be
    thrown.
 */
class CapStream : public Stream
{
protected:
  StreamPtr src;
  size_t total, pos;

 public:
  CapStream(StreamPtr _src, size_t _size)
    : src(_src), total(_size), pos(0)
  {
    assert(src->isReadable);

    isReadable = true;
    isSeekable = false;
    isWritable = false;
    hasSize = true;
    hasPosition = true;
    hasPtr = false;
  }

  size_t read(void *buf, size_t count)
  {
    if(count + pos > total)
      count = total-pos;

    size_t r = src->read(buf, count);
    pos += count;

    if(r != count || // Not enough bytes were read, OR:
       (src->eof() && !eof())) // Source eof'ed without us.
      throw std::runtime_error("CapStream: to little data in source stream");

    return count;
  }

  size_t tell() const { return pos; }
  size_t size() const { return total; }
  bool eof() const { return pos >= total; }
};

typedef boost::shared_ptr<CapStream> CapStreamPtr;

}} // namespaces
#endif
