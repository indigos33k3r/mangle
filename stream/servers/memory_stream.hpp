#ifndef MANGLE_STREAM_MEMSERVER_H
#define MANGLE_STREAM_MEMSERVER_H

#include <assert.h>
#include "../stream.hpp"
#include <string.h>

namespace Mangle {
namespace Stream {

// Do this before the class declaration, since the class itself
// uses it.
class MemoryStream;
typedef boost::shared_ptr<MemoryStream> MemoryStreamPtr;

/** A Stream wrapping a memory buffer

    This will create a fully seekable stream out of any pointer/length
    pair you give it.
 */
class MemoryStream : public Stream
{
  const void *data;
  size_t length, pos;

 public:
 MemoryStream(const void *ptr, size_t len)
   : data(ptr), length(len), pos(0)
  {
    isReadable = true;
    isSeekable = true;
    hasPosition = true;
    hasSize = true;
    hasPtr = true;
  }

  static MemoryStreamPtr Open(const void *ptr, size_t len)
  { return MemoryStreamPtr(new MemoryStream(ptr, len)); }

  MemoryStream()
    : data(NULL), length(0), pos(0)
  {
    isReadable = true;
    isSeekable = true;
    hasPosition = true;
    hasSize = true;
    hasPtr = true;
  }

  size_t read(void *buf, size_t count)
  {
    assert(data != NULL);
    assert(pos <= length);

    // Don't read more than we have
    if(count > (length - pos))
      count = length - pos;

    // Copy data
    if(count)
      memcpy(buf, ((char*)data)+pos, count);

    // aaand remember to increase the count
    pos += count;

    return count;
  }

  void seek(size_t _pos)
  {
    pos = _pos;
    if(pos > length)
      pos = length;
  }

  size_t tell() const { return pos; }
  size_t size() const { return length; }
  bool eof() const { return pos == length; }

  const void *getPtr() { return data; }
  const void *getPtr(size_t size)
    {
      // This variant of getPtr must move the position pointer
      size_t opos = pos;
      pos += size;
      if(pos > length) pos = length;
      return ((char*)data)+opos;
    }
  const void *getPtr(size_t pos, size_t size)
    {
      if(pos > length) pos = length;
      return ((char*)data)+pos;
    }

  // New members in MemoryStream:

  /// Set a new buffer and length. This will rewind the position to zero.
  void set(const void* _data, size_t _length)
  {
    data = _data;
    length = _length;
    pos = 0;
  }

  /// Clone this memory stream
  /** Make a new stream of the same buffer. If setPos is true, we also
      set the clone's position to be the same as ours.

      No memory is copied during this operation, the new stream is
      just another 'view' into the same shared memory buffer.
  */
  MemoryStreamPtr clone(bool setPos=false) const
  {
    MemoryStreamPtr res(new MemoryStream(data, length));
    if(setPos) res->seek(pos);
    return res;
  }
};

}} // namespaces
#endif
