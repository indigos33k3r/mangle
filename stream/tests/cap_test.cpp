#include <iostream>
#include <string.h>

#include "../filters/cap_stream.hpp"
#include "../servers/memory_stream.hpp"

using namespace Mangle::Stream;
using namespace std;

void test(StreamPtr inp)
{
  cout << "Size: " << inp->size() << endl;
  cout << "Pos: " << inp->tell() << endl;
  cout << "Eof: " << inp->eof() << endl;
  char data[6];

  while(!inp->eof())
    {
      memset(data, 0, 6);
      cout << "\nReading " << inp->read(data, 5) << " bytes\n";
      cout << "Result: '" << data << "'\n";
      cout << "Pos: " << inp->tell() << endl;
      cout << "Eof: " << inp->eof() << endl;
    }
}

int main()
{
  StreamPtr orig(new MemoryStream("this is a test string!", 22));
  cout << "\nEntire stream:\n--------\n";
  test(orig);

  orig->seek(4);
  StreamPtr cap(new CapStream(orig,12));
  cout << "\nCapped stream:\n--------\n";
  test(cap);

  return 0;
}
