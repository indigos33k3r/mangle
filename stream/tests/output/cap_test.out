
Entire stream:
--------
Size: 22
Pos: 0
Eof: 0

Reading 5 bytes
Result: 'this '
Pos: 5
Eof: 0

Reading 5 bytes
Result: 'is a '
Pos: 10
Eof: 0

Reading 5 bytes
Result: 'test '
Pos: 15
Eof: 0

Reading 5 bytes
Result: 'strin'
Pos: 20
Eof: 0

Reading 2 bytes
Result: 'g!'
Pos: 22
Eof: 1

Capped stream:
--------
Size: 12
Pos: 0
Eof: 0

Reading 5 bytes
Result: ' is a'
Pos: 5
Eof: 0

Reading 5 bytes
Result: ' test'
Pos: 10
Eof: 0

Reading 2 bytes
Result: ' s'
Pos: 12
Eof: 1
